### PR TITLE
Fixed definitions of const iterator and reference types

### PR DESCRIPTION
**Pull request by fuchsto**
Addresses #290 and #283

### CODE REVIEW

**Comment by fuchsto** [comment]
The positive effects of the changes in this PR are visible e.g. in examples/ex.11.halo-stencil/main.cpp:

Previously, this compiled but should lead to compile error (const-ness of parameter `data`):

```c++
void write_pgm(const std::string & filename, const Array_t & data) {
  // ...
  auto & first = data.begin();
  dash::copy(first+ext_x*y, first+ext_x*(y+1), buffer.data());
  // ...
}
```
This now leads to a compile error.
The following corrected implementation compiles without error:
```c++
void write_pgm(const std::string & filename, const Array_t & data) {
  // ...
  const auto & first = data.begin();
  dash::copy(first+ext_x*y, first+ext_x*(y+1), buffer.data());
  // ...
}
```

**Comment by fuchsto** [comment]
Includes and fixes changes from PR #284 

@fmoessbauer :

- I appreciate the ECpp item 3 quirk, but conversion between `LocalMatrixRef<T>` and `LocalMatrixRef<const T>` does not work in this case. We should find a way to make it work, though.
- Please review copy / assignment of `dash::Atomic`

File: dash/include/dash/Array.h [hunk 1]
```diff
@@ -313,7 +316,7 @@ class AsyncArrayRef
    *
    * TODO: Should return GlobAsyncPtr<...>(_array->begin())
    */
-  inline pointer begin() noexcept {
+  inline iterator begin() noexcept {
```
Comment (fmoessbauer) [inline dash/include/dash/Array.h]: `constexpr`

**Review by fmoessbauer — APPROVED** [review]
Looks good so far. These changes are really important. I created also a [code-coverage report](https://felixmoessbauer.de/temp/coverage/) for this branch.

Some tests do not succeed in coverage mode, so these are disabled: `$GTEST_FILTER="-TeamLocalityTest*:ThreadsafetyTest*"`

**Pull request merged**